### PR TITLE
Fresh results

### LINKED ISSUE #33: Get traces returns older results
Result list is currently based on the order KeyValueStore `range` API is returning. Even though this API promises no sort order, usually it returns from oldest to newest based on best effort.

We should find a way to build a result list with latest values instead of oldest.

### PR DESCRIPTION
**Pull request by jeqo**
Fix #33 

This PR includes an attempt to solve the "return oldest" issue by creating a bucketed range query procedure and validate on every bucket if we have enough traces to return a result, instead of testing every trace on request range, or return first results.

### CODE REVIEW

File: storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java [hunk 1]
```diff
@@ -178,53 +181,74 @@
...
-      spanIds.forEachRemaining(keyValue -> {
-        for (String traceId : keyValue.value) {
-          if (!traceIds.contains(traceId) && traces.size() < queryRequest.limit()) {
-            List<Span> spans = tracesStore.get(traceId);
-            if (spans != null && queryRequest.test(spans)) { // apply filters
-              traceIds.add(traceId); // adding to check if we have already add it later
-              traces.add(spans);
+      long from = (request.endTs() - request.lookback()) * 1000;
+      long to = request.endTs() * 1000;
+      long checkpoint = to - (30 * 1000 * 1000); // 30 sec before upper bound
```
Comment (jeqo) [inline storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java]: this checkpoint represent the bucket range

File: storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java [hunk 2]
```diff
@@ -178,53 +181,74 @@
...
+            });
           }
+          to = checkpoint;
+          checkpoint = checkpoint - (60 * 1000 * 1000); // 1 min before more
         }
-      });
-      LOG.debug("Traces found from query {}: {}", queryRequest, traces.size());
-      return traces;
+      }
+      traces.sort(Comparator.<List<Span>>comparingLong(o -> o.get(0).timestampAsLong()).reversed());
```
Comment (jeqo) [inline storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java]: Resulting traces list should have values close to `limit`, so sorting should not be expensive.

File: storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java [hunk 3]
```diff
@@ -111,10 +112,12 @@
 
     @Override public List<String> query() {
       List<String> serviceNames = new ArrayList<>();
-      serviceStore.all().forEachRemaining(keyValue -> {
-        // double check service names are unique
-        if (!serviceNames.contains(keyValue.value)) serviceNames.add(keyValue.value);
-      });
+      try (KeyValueIterator<String, String> all = serviceStore.all()) {
```
Comment (adriancole) [inline storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java]: sheesh do we allow dupes? sounds like we should tighten up that javadoc

File: storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java [hunk 4]
```diff
@@ -178,53 +181,74 @@
...
+                if (spans != null && request.test(spans)) { // apply filters
+                  traceIds.add(traceId); // adding to check if we have already add it later
+                  traces.add(spans);
+                }
+              }
             }
+          });
+        }
+      } else {
+        while (checkpoint > from && traces.size() < request.limit()) {
```
Comment (adriancole) [inline storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java]: cc @llinder @michaelsembwever this seems familiar though in cassandra I think we try to do async lazy chained calls.. not sure if the `traceIdsByTsStore.range` here is a remote call or not.

File: storage/src/test/java/zipkin2/storage/kafka/KafkaStorageIT.java [hunk 5]
```diff
@@ -189,6 +194,25 @@
...
+                    .execute();
+          } catch (InvalidStateStoreException e) { // ignoring state issues
+            System.err.println(e.getMessage());
+          } catch (Exception e) {
+            e.printStackTrace();
+          }
+          return traces.size() == 1
+              && traces.get(0).size() == 1; // last trace is returned first
+        });
     await().atMost(5, TimeUnit.SECONDS)
```
Comment (adriancole) [inline storage/src/test/java/zipkin2/storage/kafka/KafkaStorageIT.java]: side question.. why all the waits? I know in some cases you can't get a callback of storage quorum met. in such case you could add a sleep to a command that wraps storage to declutter the IT's main code..

File: storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java (hunk 3 above)
Comment (jeqo) [reply storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java]: yeap, that surprise me on one of my deployments where I started to get duplicate service names. Will create an issue to double check this.

Comment (jeqo) [reply storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java]: #35

File: storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java (hunk 4 above)
Comment (jeqo) [reply storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java]: All stores are local tho

File: storage/src/test/java/zipkin2/storage/kafka/KafkaStorageIT.java (hunk 5 above)
Comment (jeqo) [reply storage/src/test/java/zipkin2/storage/kafka/KafkaStorageIT.java]: Good question, probably copy-pasta heh. I think only first one is needed, will clean up.

File: storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java (hunk 4 above)
Comment (jeqo) [reply storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java]: Unless we'd support https://github.com/openzipkin/zipkin/issues/2784 where calls could span multiple instances

**Comment by jeqo** [comment]
@adriancole , I'd love to move it to oz-contrib!

File: storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java (hunk 4 above)
Comment (jeqo) [reply storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java]: Issue moved to #36

**Pull request merged**